### DISCUSSION TITLE
how to make a complex join work with filters

### DISCUSSION THREAD

**Discussion by fergusmeiklejohn** [supabase/supabase#2439] [Questions] [answered]
I'm trying use this query, but it's returning null:
```javascript
  const { data, error } = await supabase
    .from("events")
    .select(
      `*,
      organisation(
        name,
        logo,
        summary,
        social_networks(
          twitter,
          facebook,
          linkedin,
          line,
          weibo,
          tiktok,
          instagram,
          github,
          email,
          website,
          whatsapp)),
      social_networks("*")`
    )
    .eq("event_id", eventId);
```

This works: 
```javascript
  const { data, error } = await supabase
    .from("events")
    .select(
      `*,
      organisation("*"),
      social_networks("*")`
    )
    .eq("event_id", eventId);
```
But the `social_networks` table has a lot of randomly null columns, because users only have a few social networks and have different ones, so I'd like to filter out the null columns.

This also works:
```javascript
    .select(
      `*,
      organisation(
        name,
        logo,
        summary,
        social_networks),
      social_networks("*")`
    )
```
But this doesn't, and I wonder why??
```javascript
    .select(
      `*,
      organisation(
        name,
        logo,
        summary,
        social_networks("*")),
      social_networks("*")`
    )
```

Can anyone please help with how to make this type of query work? Thanks!

**Comment by steve-chavez** [comment, marked as answer]
> I'm trying use this query, but it's returning null

Could you try doing `console.log(error)`? That might reveal more details.

It seems to me that you might be hitting a bad syntax error, try removing the double quotes on `"*"`.

It's also possible that there's an invalid embed there, definitely print the error, that will clarify things.

**Reply by fergusmeiklejohn** [reply]
thanks @steve-chavez :-)
This is the error:
```bash
error accessing database:  {
  hint: "By following the 'details' key, disambiguate the request by changing the url to /origin?select=relationship(*) or /origin?select=target!relationship(*)",
  details: [
    {
      relationship: 'organisations_social_networks_fkey[social_networks][social_id]',
      origin: 'public.organisations',
      cardinality: 'm2o',
      target: 'public.social_networks'
    },
    {
      relationship: 'public.events[event_organisation_fkey][event_social_networks_fkey]',
      origin: 'public.organisations',
      cardinality: 'm2m',
      target: 'public.social_networks'
    },
    {
      relationship: 'public.speakers[speakers_organisation_fkey][speakers_social_networks_fkey]',
      origin: 'public.organisations',
      cardinality: 'm2m',
      target: 'public.social_networks'
    },
    {
      relationship: 'public.user_profiles[user_profiles_organisation_fkey][user_profiles_social_networks_fkey]',
      origin: 'public.organisations',
      cardinality: 'm2m',
      target: 'public.social_networks'
    }
  ],
  message: 'More than one relationship was found for organisations and social_networks'
}
```

**Reply by fergusmeiklejohn** [reply]
I wonder if it's because I'm referencing `social_networks` from both event and organisation?

**Reply by steve-chavez** [reply]
@fergusmeiklejohn It should work if you do it like this:
```js
    .select(
      `*,
      organisation(
        name,
        logo,
        summary,
        social_networks!social_id(*)),
      social_networks(*)`
    )
```
(Note the `social_networks!social_id(*)`)

> I wonder if it's because I'm referencing social_networks from both event and organisation?

The main problem is that besides the many-to-one(m2o) relationship from `organizations` to `social_networks`, we also detect many-to-many(m2m) relationships using other tables as junctions. But the above snippet should clear it.

**Reply by fergusmeiklejohn** [reply]
thanks @steve-chavez 
That query returns null.
```bash
error accessing database:  {
  hint: "By following the 'details' key, disambiguate the request by changing the url to /origin?select=relationship(*) or /origin?select=target!relationship(*)",
  details: [
    {
      relationship: 'organisations_social_networks_fkey[social_networks][social_id]',
      origin: 'public.organisations',
      cardinality: 'm2o',
      target: 'public.social_networks'
    },
    {
      relationship: 'public.events[event_organisation_fkey][event_social_networks_fkey]',
      origin: 'public.organisations',
      cardinality: 'm2m',
      target: 'public.social_networks'
    },
    {
      relationship: 'public.speakers[speakers_organisation_fkey][speakers_social_networks_fkey]',
      origin: 'public.organisations',
      cardinality: 'm2m',
      target: 'public.social_networks'
    },
    {
      relationship: 'public.user_profiles[user_profiles_organisation_fkey][user_profiles_social_networks_fkey]',
      origin: 'public.organisations',
      cardinality: 'm2m',
      target: 'public.social_networks'
    }
  ],
  message: 'More than one relationship was found for organisations and social_networks'
}
```

 This sql query works:
```sql
select *
from events
         join organisations org on org.org_id = events.organisation
         join social_networks org_social on org_social.social_id = org.social_networks
         join social_networks event_social on event_social.social_id = events.social_networks
where events.event_id = [EVENTID];
```

But then everything is flat I don't get the nice json nesting I get with the supabase-js client. How is that done BTW? It's so useful. It must be translating the js to sql, executing it, then parsing the response to create nested json objects?

**Reply by steve-chavez** [reply]
@fergusmeiklejohn Sorry for the late reply. 

> How is that done BTW? It's so useful. It must be translating the js to sql, executing it, then parsing the response to create nested json objects?

This is done through [PostgREST resource embedding](https://postgrest.org/en/stable/api.html#resource-embedding), which Supabase uses and `supabase-js` exposes through JS.

Btw, taking a closer look, I think your query can be solved like this:

```js
    .select(
      `*,
      organisation(
        name,
        logo,
        summary,
        social_networks!organisations_social_networks_fkey(*)),
      social_networks(*)`
    )
```

For more details, see [PostgREST embedding disambiguation](https://postgrest.org/en/stable/api.html#embedding-disambiguation).

**Reply by fergusmeiklejohn** [reply]
Thanks very much Steve!
I was stuck so I went with creating views for complex queries. The advantage there is that I can use Datagrip to iterate quickly on the local supabase instance, so I can fail fast and figure it out.
That brings the data in flat so I name columns with dots: `parent.child` then run it through unflatten to create nested objects. https://www.npmjs.com/package/flat
I also use groupBy sometimes to merge multiple rows which really all belong in the same object. Like if I'm fetching comments on posts
